### ISSUE TITLE
"Puppets"

### ISSUE THREAD

**Issue by jugendhacker** [42wim/matterbridge#788]
**Is your feature request related to a problem? Please describe.**
Currently there is only one Account relaying messages and that works, but it's not really nice for mentions etc...

**Describe the solution you'd like**
At least XMPP and Matrix support so called "puppets", where the bridge could spawn a user in the room for every user from the other side of the gateway. It let it more feel like a real guy on the same protocol.

**Comment by 42wim** [comment]
Could you give a pointer to the documentation about xmpp and matrix supporting this?

**Comment by jugendhacker** [comment]
I'm not so into Matrix so I don't know how it works there...

But in XMPP you will need some [External Component](https://xmpp.org/extensions/xep-0114.html) to implement this... but could be much work

**Comment by poVoq** [comment]
There is a go based bridge that does this on IRC (from Discord):
https://github.com/qaisjp/go-discord-irc
The author even expressed interest in adding this feature to Matterbridge @qaisjp

**Comment by qaisjp** [comment]
lol hi, see #667, i still plan to work on this _eventually_

**Comment by poVoq** [comment]
https://github.com/jlu5/PyLink

Another example of a nicely working IRC puppeteering bridge

**Comment by LucesEnBabel** [comment]
It would be great if we can get this on Matrix!

**Comment by jugendhacker** [comment]
@LucesEnBabel it is great in every protocol where it is theoretically possible....

**Comment by qaisjp** [comment]
I've been looking into this, and I'm not sure what's the best way to go about this config-wise.

**Example base config to work off**

Take this simple bi-directional configuration that bridges between Discord and IRC.

```
[discord]
[discord.compsoc]
Token="<some value>"
Server="<some value>"
WebhookURL="<some value>"

[irc]
[irc.compsoc]
Server="localhost:6697"
Nick="DiscordBot"

[[gateway]]
name="compsoc.bottest"
enable=true

[[gateway.inout]]
account="irc.compsoc"
channel="#bottest"

[[gateway.inout]]
account="discord.compsoc"
channel="bottest"
```

**Naming**

Discord already support puppets, but they are... "ephemeral". We don't want to people to use a `Puppets` setting and realise they actually need to use `WebhookURL`. That can be considered poor user experience.

We know that Discord puppets are _not_ persistent, unlike IRC. So we could name our setting something like `PersistentPuppets`.

**Flow of knowledge**

User information flows from `discord` to `irc`. How do we represent this information in the config?

And we probably want a configuration design that can gracefully support more than just IRC puppets.

**Suggestion**

What if we did something like this?

```
[[gateway]]
name="compsoc.bottest"
enable=true

[[gateway.inout]]
account="irc.compsoc"
channel="#bottest"
puppeting="in"

[[gateway.inout]]
account="discord.compsoc"
channel="bottest"
puppeting="out"
```

I would actually suggest something like this, but I think including the term [`bridge`](https://github.com/42wim/matterbridge/blob/83ca8777b9ce99026c2deaa6d32209d9e2c2abf1/bridge/config/config.go#L166-L172) might confuse people that are writing their config:

```
[[gateway]]
name="compsoc.bottest"
enable=true

[[gateway.bridge]]
account="irc.compsoc"
channel="#bottest"
chat="inout"
users="out"   # output

[[gateway.bridge]]
account="discord.compsoc"
channel="bottest"
chat="inout"
users="in"    # input
```

----

What do you think? How should this be designed?

**Comment by poVoq** [comment]
Proposal to implement it via webooks in XMPP (Prosody only):
https://github.com/42wim/matterbridge/issues/1082

**Comment by jugendhacker** [comment]
Not that nice, because it's a non standard thing...

**Comment by poVoq** [comment]
Yeah, but it reuses existing already implemented code and can be a temporary solution. Because let's be realistic, a full xmpp component isn't coming anytime soon.

**Comment by kousu** [comment]
There's no need for webhooks to do do puppetting in xmpp. Because of a quirk of how xmpp was designed, a single xmpp connection can join a groupchat any number of times under any number of names. You can try this out yourself: just sign on in two apps then join the same groupchats in them, but pick different nicknames when you do: you'll show up as two people. Some apps will even happily let you join the same room multiple times under multiple names (not Conversations though, which is probably for the best).

To do this, matterbridge just needs to be sending stanzas like this:

```
<presence from='bridge@example.com' to='room@conference.example.com/alice' />
```

```
<message from='room@conference.example.com/alice' to='room@conference.example.com/'>
  <body> ..... </body>
</message>
```

to spoof a different user, just do


```
<message from='room@conference.example.com/bob' to='room@conference.example.com/'>
  <body> ..... </body>
</message>
```

The full spec is here: https://xmpp.org/extensions/xep-0045.html

I'm not 100% sure about this, but I don't think you even need to send the initial presence for the messages to go through. Presence is fairly optional in xmpp.

**Comment by jugendhacker** [comment]
@kousu first the presence is not optional in MUCs and you need to send it for every Nick you want because that is how you join a room in XMPP and usually you can't write in a MUC if you aren't joined.... the server will throw an error for this.

Second: I also tough about going the way with one JID and multiple nicks but it's actually not the right way to do this... you wan't to do this via a component (https://xmpp.org/extensions/xep-0114.html) and a JID at this component for every user. But this will definitely need server support because you have to configure components on the XMPP Server.

**Comment by kousu** [comment]
Why not? Everything matterbridge does is kind of kludgy (but very effective) glue to get systems not meant to talk to each other to talk. I don't see how there can be a right or a wrong way to make this happen.

**Comment by jugendhacker** [comment]
Yeah it is the simplest option....

**Comment by kousu** [comment]
The external component protocol is good for adding individuals to your contact list who are on other protocols, for example having xmpp:friend1@facebook.component.example.com and xmpp:friend2@facebook.component.example.com, or for connecting from XMPP to a foreign protocol (e.g.  for example the  [biboumi](https://biboumi.louiz.org) component at xmpp:irc.hackint.org will let you connect to IRC rooms on irc.hackint.org from XMPP). But matterbridge explicitly is not about personal rosters or DMs, it's only about groupchats.

**Comment by jugendhacker** [comment]
No the component protocol is not just for that... it could basically used for everything because you could implement like a second server behind a server without needing to care about the routing and stuff. But I agree with you that this might be overkill for matterbridge.

**Comment by kousu** [comment]
> @kousu first the presence is not optional in MUCs and you need to send it for every Nick you want because that is how you join a room in XMPP and usually you can't write in a MUC if you aren't joined.... the server will throw an error for this.

I just confirmed, you're right, https://xmpp.org/extensions/xep-0045.html#enter-muc says

> In order to participate in the discussions held in a multi-user chat room, a user MUST first become an occupant by entering the room.
> MUC clients MUST signal their ability to speak the MUC protocol by including in the initial presence stanza an empty <x/> element ...

but anyway, that isn't a big deal. It just means that matterbridge needs to hold onto a little bit of state, a flag for whether each username has sent presence or not.

**Comment by Francesco149** [comment]
I know this is old but +1, it would be great to have puppeting

**Comment by poVoq** [comment]
For Matrix there is open PR for this now, and on modern IRC servers like Ergo.chat the user spoofing via RELAYMSG works pretty well. But I agree a better solution for XMPP would be nice, ideally via Component support similar to the proposed appservice implementation for Matrix.
